test/e2e/run/e2e.go

```diff
@@ -0,0 +1,264 @@
+	Configs         string
+	Ginkgo          string
+	GinkgoOutputLog string
+	Reports         string
+	ReportsFileJSON string
+	SetupLog        string
+	// prefix (path) for logs/artifacts on S3, instance name will be appended to this path
+	// ex: logs/<cluster-name>
+	LogsBucketPath string
+	// either e2e.test or ./test/e2e/suite
+	TestsBinaryOrSource string
+	// path to test config file e2e.TestConfig
+	TestConfigFile string
+	// path to test resources file cluster.TestResources
+	TestResourcesFile string
+}
+
+type E2E struct {
+	testPhases      []Phase
+	AwsCfg          aws.Config
+	Logger          logr.Logger
+	NoColor         bool
+	Paths           E2EPaths
+	TestConfig      e2e.TestConfig
+	TestLabelFilter string
+	TestProcs       int
+	TestTimeout     string
+	TestResources   cluster.TestResources
+	SkipCleanup     bool
+	SkippedTests    string
+}
+
+func (e *E2E) Run(ctx context.Context) (E2EResult, error) {
+	e.initPaths()
+
+	err := e.preTestSetup()
+	if err != nil {
+		return E2EResult{Phases: e.testPhases}, err
+	}
+
+	// After this point we are going to run as much as we can regardeless of potential errors
+	// and will collect all errors to return them at the end
+	allErrors := []error{}
+
+	// if err := e.runTests(ctx); err != nil {
+	// 	allErrors = append(allErrors, err)
+	// }
+
+	e2eResult, err := e.parseReport(ctx)
+	if err != nil {
+		allErrors = append(allErrors, err)
+	}
+
+	if err := e.uploadArtifactsToS3(ctx); err != nil {
+		allErrors = append(allErrors, err)
+	}
+
+	e2eResult.Phases = e.testPhases
+	return e2eResult, errors.Join(allErrors...)
+}
+
+func (e *E2E) initPaths() {
+	e.Paths.Configs = filepath.Join(e.TestConfig.ArtifactsFolder, e2eConfigFolder)
+	e.Paths.TestConfigFile = filepath.Join(e.Paths.Configs, e2eConfigFile)
+	e.Paths.LogsBucketPath = filepath.Join(constants.TestS3LogsFolder, e.TestConfig.ClusterName)
+	e.Paths.Reports = filepath.Join(e.TestConfig.ArtifactsFolder, e2eReportsFolder)
+	e.Paths.ReportsFileJSON = filepath.Join(e.Paths.Reports, e2eReportsFileJSON)
+	e.Paths.TestResourcesFile = filepath.Join(e.Paths.Configs, e2eTestResourcesFile)
+	e.Paths.GinkgoOutputLog = filepath.Join(e.TestConfig.ArtifactsFolder, testGinkgoOutputLog)
+	e.Paths.CleanupLog = filepath.Join(e.TestConfig.ArtifactsFolder, testCleanupLogFile)
+	e.Paths.SetupLog = filepath.Join(e.TestConfig.ArtifactsFolder, testSetupLogFile)
+}
+
+func (e *E2E) addTestPhase(name, message string, err error) Phase {
+	phase := Phase{Name: name, Status: "success"}
+	if err != nil {
+		phase.Error = fmt.Errorf(message, err)
+		phase.FailureMessage = err.Error()
+		phase.Status = "failure"
+	}
+	e.testPhases = append(e.testPhases, phase)
+	return phase
+}
+
+// preTestSetup sets up the directories and writes the test configs
+// it returns the failed phase and an error if one occurred
+func (e *E2E) preTestSetup() error {
+	err := e.setupDirectories()
+	dirPhase := e.addTestPhase(phaseNameSetupDirectories, "setting up directories: %w", err)
```

Comment:
I find passing formatting strings a bit confusing to read and error prone (is not part of the method signature and can't be enforced at compile time
What about just doing the wrapping in addTestPhase?